crates/currency/src/lib.rs

```diff
@@ -0,0 +1,326 @@
+//! # Currency Module
+//! Based on the [Collateral specification](https://interlay.gitlab.io/polkabtc-spec/spec/collateral.html).
+//! Based on the [Treasury specification](https://interlay.gitlab.io/polkabtc-spec/spec/treasury.html).
+
+#![deny(warnings)]
+#![cfg_attr(test, feature(proc_macro_hygiene))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+use frame_support::{
+    dispatch::{DispatchError, DispatchResult},
+    ensure,
+    traits::{Currency, ExistenceRequirement, ReservableCurrency},
+};
+
+pub type BalanceOf<T, I = ()> =
+    <<T as Config<I>>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+pub use pallet::*;
+
+pub type Collateral = pallet::Instance1;
+pub type Treasury = pallet::Instance2;
+
+pub const BTC_DECIMALS: u8 = 8;
+pub const DOT_DECIMALS: u8 = 10;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use frame_support::pallet_prelude::*;
+
+    /// ## Configuration
+    /// The pallet's configuration trait.
+    #[pallet::config]
+    pub trait Config<I: 'static = ()>: frame_system::Config {
+        /// The asset to wrap
+        type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+
+        /// The overarching event type.
+        type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
+    }
+
+    // The pallet's events
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    #[pallet::metadata(T::AccountId = "AccountId", BalanceOf<T, I> = "Balance")]
+    pub enum Event<T: Config<I>, I: 'static = ()> {
+        Mint(T::AccountId, BalanceOf<T, I>),
+        Lock(T::AccountId, BalanceOf<T, I>),
+        Unlock(T::AccountId, BalanceOf<T, I>),
+        Burn(T::AccountId, BalanceOf<T, I>),
+        Release(T::AccountId, BalanceOf<T, I>),
+        Slash(T::AccountId, T::AccountId, BalanceOf<T, I>),
+    }
+
+    #[pallet::error]
+    pub enum Error<T, I = ()> {
+        /// Account has insufficient free balance
+        InsufficientFreeBalance,
+        /// Account has insufficient reserved balance
+        InsufficientReservedBalance,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config<I>, I: 'static> Hooks<T::BlockNumber> for Pallet<T, I> {}
+
+    /// Note that an account's free and reserved balances are handled
+    /// through the Balances module.
+    ///
+    /// Total locked balance
+    #[pallet::storage]
+    pub type TotalLocked<T: Config<I>, I: 'static = ()> = StorageValue<_, BalanceOf<T, I>, ValueQuery>;
+
+    /// The number of decimals used to represent one unit.
+    #[pallet::storage]
+    #[pallet::getter(fn decimals)]
+    pub type Decimals<T: Config<I>, I: 'static = ()> = StorageValue<_, u8, ValueQuery>;
+
+    #[pallet::genesis_config]
+    pub struct GenesisConfig<T: Config<I>, I: 'static = ()> {
+        pub marker: sp_std::marker::PhantomData<(T, I)>,
```

Comment:
What is the marker?

Reply:
None of the fields need to refer to the trait so we use [PhantomData](https://doc.rust-lang.org/std/marker/struct.PhantomData.html) to mark this (otherwise it will not compile).

Reply:
Thanks!